cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -39,12 +43,330 @@
+    save_key_data_.resize(0);
+    return Status::OK();
+  }
+
+  // Checks whether the given grouping data extends the current segment, i.e., is equal to
+  // previously seen grouping data, which is updated with each invocation.
+  bool Extend(const void* data) {
+    size_t byte_width = static_cast<size_t>(key_type_.type->byte_width());
+    bool extends = save_key_data_.size() != byte_width
+                       ? kDefaultExtends
+                       : 0 == memcmp(save_key_data_.data(), data, byte_width);
+    save_key_data_.resize(byte_width);
+    memcpy(save_key_data_.data(), data, byte_width);
+    return extends;
+  }
+
+  Result<Segment> GetNextSegment(const Scalar& scalar, int64_t offset, int64_t length) {
+    ARROW_RETURN_NOT_OK(CheckType(*scalar.type));
+    if (!scalar.is_valid) {
+      return Status::Invalid("segmenting an invalid scalar");
+    }
+    auto data = checked_cast<const PrimitiveScalarBase&>(scalar).data();
+    bool extends = length > 0 ? Extend(data) : kEmptyExtends;
+    return MakeSegment(length, offset, length, extends);
+  }
+
+  Result<Segment> GetNextSegment(const DataType& array_type, const uint8_t* array_bytes,
+                                 int64_t offset, int64_t length) {
+    RETURN_NOT_OK(CheckType(array_type));
+    int64_t byte_width = array_type.byte_width();
+    int64_t match_length = GetMatchLength(array_bytes + offset * byte_width, byte_width,
+                                          array_bytes, offset, length);
+    bool extends = length > 0 ? Extend(array_bytes + offset * byte_width) : kEmptyExtends;
+    return MakeSegment(length, offset, match_length, extends);
+  }
+
+  Result<Segment> GetNextSegment(const ExecSpan& batch, int64_t offset) override {
+    ARROW_RETURN_NOT_OK(CheckForGetNextSegment(batch, offset, {key_type_}));
+    if (offset == batch.length) {
+      return MakeSegment(batch.length, offset, 0, kEmptyExtends);
+    }
+    const auto& value = batch.values[0];
+    if (value.is_scalar()) {
+      return GetNextSegment(*value.scalar, offset, batch.length);
+    }
+    ARROW_DCHECK(value.is_array());
+    const auto& array = value.array;
+    if (array.GetNullCount() > 0) {
+      return Status::NotImplemented("segmenting a nullable array");
+    }
+    return GetNextSegment(*array.type, GetValuesAsBytes(array), offset, batch.length);
+  }
+
+ private:
+  TypeHolder key_type_;
+  std::vector<uint8_t> save_key_data_;
+};
+
+struct AnyKeysSegmenter : public BaseRowSegmenter {
+  static Result<std::unique_ptr<RowSegmenter>> Make(
+      const std::vector<TypeHolder>& key_types, ExecContext* ctx) {
+    ARROW_RETURN_NOT_OK(Grouper::Make(key_types, ctx));  // check types
+    return std::make_unique<AnyKeysSegmenter>(key_types, ctx);
+  }
+
+  AnyKeysSegmenter(const std::vector<TypeHolder>& key_types, ExecContext* ctx)
+      : BaseRowSegmenter(key_types),
+        ctx_(ctx),
+        grouper_(nullptr),
+        save_group_id_(kNoGroupId) {}
+
+  Status Reset() override {
```

Comment:
Is this used anywhere? When do we want to reset a segmenter?

Reply:
No, it is not used, and still it is useful to be able to reset a segmenter. It looks like we disagree on what justifies adding code. I feel your approach is to remove all code that is not currently used, considering a particular application in mind. My approach is different, accounting for library usage, and I'll opt to add such code when I find other justifications for doing so, like that it is simple/safe to implement or useful for future authoring, especially in the more generic parts of the code. I'd keep the code here.

Reply:
How would this be used in the future? Also, how is this tested if currently not used?

Reply:
I won't hold it if you feel strongly. But when I often prefer to keep code changes to minimal (especially considering this is a already a 1000+ line change set) plus as a reviewer, the intent of the author of keeping unused method for future is unclear to me. So if you feel strongly about keeping those I'd suggesting explaining how u think this can be used in the future. (It's not clear to me when we would want to reset a segmenter)

Reply:
I'll add test coverage for segmenter resetting.

> In your mind, how would this be used in the future?

For example, outside the context of this PR, it makes sense to reset a segmenter when each batch is processed independently, rather than within a stream of batches, so that a segment must not cross into the next batch.

Reply:
Ok

Reply:
We should also keep in mind that "custom exec node developers" are a valid persona to support.  As we build up utilities like these we could in theory even expose them to python (e.g. pyarrow users have asked for things like ExecBatchBuilder and the row table in the past).  I think it's ok to err a little bit on the side of "a complete abstraction" as long as it doesn't go overboard.

To help review length in the future, we could probably have reviewed the segmenters independently and then come back and reviewed the changes to the node itself.  But that ship has sailed at this point :)

Reply:
> we could probably have reviewed the segmenters independently

Perhaps, but that may have replaced one problem with another. In practice, I would have locally developed the same code as in this PR so I'd know the segmenter API is in fact what is needed for implementing segmented aggregation. Then I would have created a PR for just the segmenter code, and questions around why this or that particular code was needed would have popped up. It would be much harder to justify the segmenter code when segmented aggregation code is not shown. Moreover, even when the segmenter code were accepted, the next PR for segmented aggregation would likely have led to requested changes that affect the segmenter API too. In short, it's possible we would not have gained much from phasing this way.

OTOH, I think it can be useful to start with the larger PR and after some discussions decide to split it up, because then it is clear how the full code is going to look like and how the phases are justified.

Reply:
I actually agree with Weston that smaller PR helps in this case. Reviewing a 1k+ line PR is quite painful and very easy to miss things. Let's try breaking up into smaller PR's next time.